### PR TITLE
[ClassLoader] Fix storing not-found classes in APC cache

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 2.3 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by xabbuh** [comment]
Did this cause issues somewhere? The docs state that the return value will also be `false` on failures.

**Comment by nicolas-grekas** [comment]
Its false on failure, but also false when false is the cached value... $success is the way to check for cache hit or miss.

**Comment by xabbuh** [comment]
:+1: ah of course

Status: Reviewed

File: src/Symfony/Component/ClassLoader/ApcClassLoader.php [hunk 1]
```diff
@@ -118,11 +118,13 @@ public function loadClass($class)
      *
      * @param string $class A class name to resolve to file
      *
-     * @return string|null
+     * @return string|false
```
Comment (Tobion) [inline src/Symfony/Component/ClassLoader/ApcClassLoader.php]: This does not seem correct. The other class loaders have `string|null`. As they are decorated by ApcClassLoader something is wrong. If they do not return false at all, then this whole change seems irrelevant.

**Comment by Tobion** [comment]
Status: Needs Work

**Comment by nicolas-grekas** [comment]
@Tobion fixed
Status: need review

File: src/Symfony/Component/ClassLoader/ApcClassLoader.php [hunk 2]
```diff
@@ -122,8 +122,10 @@ public function loadClass($class)
      */
     public function findFile($class)
     {
-        if (false === $file = apcu_fetch($this->prefix.$class)) {
-            apcu_store($this->prefix.$class, $file = $this->decorated->findFile($class));
+        $file = apcu_fetch($this->prefix.$class, $success);
+
+        if (!$success) {
+            apcu_store($this->prefix.$class, $file = $this->decorated->findFile($class) ?: null);
```
Comment (Tobion) [inline src/Symfony/Component/ClassLoader/ApcClassLoader.php]: what is the point of the null condition?

Comment (nicolas-grekas) [inline src/Symfony/Component/ClassLoader/ApcClassLoader.php]: Composer's ClassLoader returns false here.

**Comment by Tobion** [comment]
`ApcUniversalClassLoader` needs the same fix.

File: src/Symfony/Component/ClassLoader/ApcClassLoader.php (hunk 2 above)
Comment (Tobion) [inline src/Symfony/Component/ClassLoader/ApcClassLoader.php]: then WinCache and Xcache class loaders should do the same normalization

**Comment by nicolas-grekas** [comment]
@Tobion PR updated

**Comment by Tobion** [comment]
Thank you @nicolas-grekas.
